src/rmc.rs

```diff
@@ -0,0 +1,302 @@
+#[derive(Clone)]
+pub enum Message<H: Signable, S: Signature, M: PartialMultisignature> {
+    SignedHash(UncheckedSigned<Indexed<H>, S>),
+    MultisignedHash(UncheckedSigned<H, M>),
+}
+
+impl<H: Signable, S: Signature, M: PartialMultisignature> Message<H, S, M> {
+    pub fn hash(&self) -> &H {
+        match self {
+            Message::SignedHash(unchecked) => unchecked.as_signable().as_signable(),
+            Message::MultisignedHash(unchecked) => unchecked.as_signable(),
+        }
+    }
+}
+
+/// Stores the current phase of a hash, and all relevant data useful in that phase:
+///
+/// * in the signature aggregation phase we store a signed hash and an aggregated partial
+/// multisignature under the hash
+/// * in the multisignature aggregation phase, we store the complete multisignature of the hash
+/// and the set of nodes about who we know that they have a multisignature
+/// * in the completed phase we only store the multisigned hash.
+enum HashState<'a, H: Signable, MK: MultiKeychain> {
+    SignatureAggregation {
+        partial: PartiallyMultisigned<'a, H, MK>,
+    },
+    Completed {
+        complete: Multisigned<'a, H, MK>,
+    },
+}
+
+impl<'a, H: Signable + Clone, MK: MultiKeychain> HashState<'a, H, MK> {
+    fn handle_unchecked_message(
+        maybe_state: Option<Self>,
+        message: Message<H, MK::Signature, MK::PartialMultisignature>,
+        keychain: &'a MK,
+        multisigned_hashes_tx: &UnboundedSender<Multisigned<'a, H, MK>>,
+    ) -> Option<Self> {
+        match maybe_state {
+            Some(HashState::Completed { .. }) => maybe_state,
+            _ => match message {
+                Message::MultisignedHash(unchecked) => match unchecked.check_multi(keychain) {
+                    Ok(complete) => Some(HashState::Completed { complete }),
+                    Err(_) => {
+                        debug!(target: "rmc", "Received a hash with a bad signature");
+                        maybe_state
+                    }
+                },
+                Message::SignedHash(unchecked) => {
+                    let signed_hash = match unchecked.check(keychain) {
+                        Ok(signed_hash) => signed_hash,
+                        Err(_) => {
+                            debug!(target: "rmc", "Received a hash with a bad multisignature");
+                            return maybe_state;
+                        }
+                    };
+                    let partial = match maybe_state {
+                        None => signed_hash.into_partially_multisigned(keychain),
+                        Some(HashState::SignatureAggregation { mut partial }) => {
+                            partial.add_signature(signed_hash);
+                            partial
+                        }
+                        Some(HashState::Completed { .. }) => unreachable!(),
+                    };
+                    match partial.try_into_complete(keychain) {
```

Comment:
I like the idea of making `PartiallyMultisigned<'a, T, MK>` an enum and pattern matcing it here instead of calling the conversion function. Will you change it?

Reply:
Done